### PR TITLE
Add hard sigmoid

### PR DESCRIPTION
**Pull request by MARD1NO**
增加 hardsigmoid 算子
**暂时没做inplace**
Pytorch实现有误区, 边界值依旧有梯度,做不了inplace操作
Paddle的边界值则没有梯度

## 概述
实现hardsigmiod算子，对标Pytorch的 torch.nn.Hardsigmoid
图片

## 功能 CheckList
**注意** : 功能复选框均为可选项，若未选择，说明理由即可。例如：该 Op 由 Python 接口拼接而成，因此无 `SetBatchAxisInferFn` Op 注册；再比如：该 Op 无输入，因此无 `SetInputArgModifyFn`。

模板中自带的复选框可留空，但是不能删除。可根据实际情况增加复选框选项。

### Op
 - [x] Op SetBatchAxisInferFn
 - [x] Op SetGetSbpFn
 - [x] Op SetInputArgModifyFn
 - [x] Op 反向梯度注册

### Kernel
注册浮点类型, 仅仅针对前向注册 int整数类型
 - [x] CPU in:float32 
 - [x] CPU in:float64

 - [x] GPU in:float32 
 - [x] GPU in:float64
 - [x] GPU in:float16

### Python Wrapper
 - [x] Python API 参数检查及异常提示
 - [x] 接口注释
 - [x] Example 

### 测试
 - [x] 单机单卡  CPU Test Case 
 - [x] 单机单卡  GPU Test Case 
 - [ ] 单机多卡  CPU Test Case 
 - [x] 单机多卡  GPU Test Case
 - [ ] 分布式  CPU Test Case  
 - [ ] 分布式  GPU Test Case  

## GPU 有效带宽
带 GPU 的 Op，请参考 https://github.com/Oneflow-Inc/OneTeam/issues/167 测试有效带宽，并附带测试报告。
以下是报告样例：

理论带宽：
```
 Device to Device Bandwidth, 1 Device(s)
 PINNED Memory Transfers
   Transfer Size (Bytes)	Bandwidth(GB/s)
   32000000			378.4
```

实际带宽：
```

I1230 17:54:02.273802 233663 kernel.cpp:89] PROFILER::KERNEL::CUDA_MEMORY_BANDWIDTH op_name: HardSigmoid_1 elapsed(ms): 0.177888 memory_size(Byte): 50331648 bandwidth(GB/s): 263.508
```
折合成百分比为69.64%

## PR Checklist
 - [ ] PR 标题语句通畅，明确表达 PR 内容，适合直接作为新版本发布时的 changelog
 - [x] 代码格式化 
 - [x] 已经本地编译通过
 - [ ] 已本地针对改动测试
 - [ ] 已添加 type 标签:(填写 type 标签名，如 `bug, enhancement, purge, feature, documentation`)
 - [ ] 已添加 component 标签:(填写 component 标签名，如 `op, system, eager, build, xla, python, ci, test, tooling, onnx`)
 - [ ] Draft 转正式 PR 前已请人 Review

### CODE REVIEW

File: oneflow/user/kernels/hardsigmoid_kernel.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+#define REGISTER_CPU_HARDSIGMOID_KERNEL(device, dtype)    \
+  REGISTER_USER_KERNEL("hardsigmoid")                     \
+      .SetCreateFn<CpuHardsigmoidKernel<device, dtype>>() \
+      .SetIsMatchedHob((HobDeviceTag() == device)         \
+                       & (HobDataType("out", 0) == GetDataType<dtype>::value));
+
+REGISTER_CPU_HARDSIGMOID_KERNEL(DeviceType::kCPU, int32_t);
+REGISTER_CPU_HARDSIGMOID_KERNEL(DeviceType::kCPU, int64_t);
+REGISTER_CPU_HARDSIGMOID_KERNEL(DeviceType::kCPU, float);
```
Comment (Ldpe2G) [inline oneflow/user/kernels/hardsigmoid_kernel.cpp:58]: 这两个类型，支持前向， 不支持后向？那感觉可以去掉了

File: oneflow/user/kernels/hardsigmoid_kernel.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+};
+
+#define REGISTER_CPU_HARDSIGMOID_KERNEL(device, dtype)    \
+  REGISTER_USER_KERNEL("hardsigmoid")                     \
+      .SetCreateFn<CpuHardsigmoidKernel<device, dtype>>() \
+      .SetIsMatchedHob((HobDeviceTag() == device)         \
+                       & (HobDataType("out", 0) == GetDataType<dtype>::value));
+
+REGISTER_CPU_HARDSIGMOID_KERNEL(DeviceType::kCPU, int32_t);
+REGISTER_CPU_HARDSIGMOID_KERNEL(DeviceType::kCPU, int64_t);
```
Comment (Ldpe2G) [inline oneflow/user/kernels/hardsigmoid_kernel.cpp:57]: 上面两个类型，支持前向， 不支持后向？那感觉可以去掉了

File: oneflow/python/test/ops/test_hardsigmoid.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        for i in range(elem_cnt):
+            if input[i] > -3 and input[i] < 3:
+                diff[i] = 1 / 6
+        diff = np.reshape(diff, newshape=input_shape)
+        return diff
+
+    _np_grad = np_diff(input_1)
+
+    def assert_prediction_grad(blob: tp.Numpy):
+        # print("NUmpy grad is: ", _np_grad)
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_hardsigmoid.py:83]: debug的注释去掉吧

File: oneflow/python/test/ops/test_hardsigmoid.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        assert np.allclose(of_out_hardsigmoid, np_out_hardsigmoid, atol=1e-5)
+
+
+def _gen_arg_dict(shape, device_type, value_type, machine_ids, device_counts):
+    # Generate a dict to pass parameter to test case
+    arg_dict = OrderedDict()
+    arg_dict["input_shape"] = [shape]
+    arg_dict["device_type"] = [device_type]
+    if value_type == "float" and device_type == "cpu":
+        arg_dict["value_type"] = [
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_hardsigmoid.py:158]: 这个 value_type  tuple 的内容是都一样的，可以考虑直接合并成一个 data_type 变量

**Review by Ldpe2G — APPROVED** [review]
